### PR TITLE
add transfer learning step

### PR DESCRIPTION
**Pull request by kordc**
This change adds transfer learning step to steps.py and introduces a workaround of saving callbacks in the state's json file.

### CODE REVIEW

File: art/steps.py [hunk 1]
```diff
@@ -570,3 +577,111 @@ def do(self, previous_states: Dict):
...
+        """Adds freezing to the model"""
+
+        def freeze_by_name(model):
+            """Freeze parameters by layer names."""
+            for name in self.freeze_names:
+                for param in model.named_parameters():
+                    if name in param[0]:
+                        param[1].requires_grad = False
+
+        def freeze_to_layer(model):
```
Comment (trebacz626) [inline art/steps.py]: OK, but I think telling it to not freeze last n layers would be more usefull

**Review by trebacz626 — DISMISSED** [review]

File: art/steps.py (hunk 1 above)
Comment (kordc) [reply art/steps.py]: Okay, I may change it.

Comment (SebChw) [reply art/steps.py]: Yes it would be easier for user to tell up to which not to freeze. Now you force user to tell up to which freeze. Probably this is harder.

**Comment by kordc** [comment]
#90

**Review by trebacz626 — DISMISSED** [review]

File: art/project.py [hunk 2]
```diff
@@ -192,10 +192,13 @@ def run_all(self, force_rerun=False):
         run_id = get_run_id()
         logger_id = add_logger(EXPERIMENT_LOG_DIR / get_new_log_file_name(run_id))
         try:
-            for step in self.steps:
-                self.metric_calculator.compile(step["skipped_metrics"])
-                step, checks = step["step"], step["checks"]
+            for step_check in self.steps:
```
Comment (SebChw) [inline art/project.py]: I'd think of different name maybe `step_dict`?

File: art/project.py [hunk 3]
```diff
@@ -192,10 +192,13 @@ def run_all(self, force_rerun=False):
...
+                self.metric_calculator.compile(step_check["skipped_metrics"])
+                step, checks = step_check["step"], step_check["checks"]
                 self.state.current_step = step
+                if isinstance(step, ModelStep):
```
Comment (SebChw) [inline art/project.py]: I think that project shouldn't be responsible for setting devices. I think perfect place to put it is inside `ModelStep.__call__` . We have access to both metric calculator and trainer
image

File: art/steps.py [hunk 4]
```diff
@@ -219,8 +219,6 @@ def __init__(
             logger (Optional[Union[Logger, Iterable[Logger], bool]], optional): Logger to be used. Defaults to None.
         """
         super().__init__()
-        if logger is not None:
```
Comment (SebChw) [inline art/steps.py]: Put this back I'll merge your PR concerning this

File: art/steps.py [hunk 5]
```diff
@@ -235,6 +233,7 @@ def __init__(
 
         self.model_name = model_class.__name__
         self.hash = self.model_class.get_hash()
+        self.trainer = Trainer(**self.trainer_kwargs, logger=self.logger)
```
Comment (SebChw) [inline art/steps.py]: You create trainer twice now  here in __init__ and in __call__

**Review by SebChw — CHANGES_REQUESTED** [review]

File: art/steps.py (hunk 5 above)
Comment (kordc) [reply art/steps.py]: Yes, because for now in every call the trainer is initialized once again. I initiatialize it because I need the torch device from it from the init, before the call

File: art/project.py (hunk 3 above)
Comment (kordc) [reply art/project.py]: Oh great, I didn't see the metric calculator there

File: art/steps.py (hunk 5 above)
Comment (kordc) [reply art/steps.py]: Okay, doesn't matter, I didn't see the first comment

**Review by SebChw — APPROVED** [review]
